os/hal/ports/STM32/LLD/I2Cv1/hal_i2c_lld.c

```diff
@@ -735,6 +735,48 @@ void i2c_lld_stop(I2CDriver *i2cp) {
   }
 }
 
+/**
+ * @brief   Disable DMA, disable interrupts, but leave the peripheral enabled
+ *
+ * @param[in] i2cp      pointer to the @p I2CDriver object
+ *
+ * @notapi
+ */
+void i2c_lld_soft_stop(I2CDriver *i2cp) {
+
+  /* If not in stopped state then disables the I2C clock.*/
+  if (i2cp->state != I2C_STOP) {
+
+    /* I2C disable.*/
+    dmaStreamFreeI(i2cp->dmatx);
+    dmaStreamFreeI(i2cp->dmarx);
+    i2cp->dmatx = NULL;
+    i2cp->dmarx = NULL;
+
+#if STM32_I2C_USE_I2C1
+    if (&I2CD1 == i2cp) {
+      nvicDisableVector(I2C1_EV_IRQn);
+      nvicDisableVector(I2C1_ER_IRQn);
+    }
+#endif
+
+#if STM32_I2C_USE_I2C2
+    if (&I2CD2 == i2cp) {
+      nvicDisableVector(I2C2_EV_IRQn);
+      nvicDisableVector(I2C2_ER_IRQn);
+    }
+#endif
+
+#if STM32_I2C_USE_I2C3
+    if (&I2CD3 == i2cp) {
+      nvicDisableVector(I2C3_EV_IRQn);
+      nvicDisableVector(I2C3_ER_IRQn);
+    }
+#endif
+  }
+}
+
+

```

Comment:
i2c_lld_soft_stop() looks like 99% same as i2c_lld_stop() excluding several removed lines.
Does it make sense to just add an option to existing function?, ex. i2c_lld_stop(..., disablePeripheral = true)

Reply:
I don't feel like changing the signature of i2c_lld_stop() as it would break all callers and become a nightmare for future updates of Chibios. What I thought about doing was to create a helper function that can be shared by the 2 functions, but looking at how this driver is implemented, it looks like duplicating the code is the most used method rather than using smaller shared functions (see all the initialization for each i2c bus)